### PR TITLE
feat(condo): DOMA-8258 add NotificationAnonymousSetting for check dis…

### PR DESCRIPTION
**Pull request by ekabardinsky**
…abled transports & message types

### CODE REVIEW

File: apps/condo/domains/notification/UnsubscribeMiddleware.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+const express = require('express')
+
+const { UNSUBSCRIBE_LINK_PATH } = require('@condo/domains/notification/constants/links')
```
Comment (pahaz) [inline apps/condo/domains/notification/UnsubscribeMiddleware.js:3]: This const used only here. It's better to move it inside this file.

File: apps/condo/domains/notification/routes/unsubscribeLinkRouter.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+const redisGuard = new RedisGuard()
+
+class UnsubscribeLinkRouter {
+    async init () {
+        const { keystone: context } = await getSchemaCtx('MultiPayment')
+        this.context = context
+    }
+
+
+    async checkLimits (req) {
```
Comment (pahaz) [inline apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:26]:
```suggestion
    }

    async checkLimits (req) {
```

File: apps/condo/domains/notification/routes/unsubscribeLinkRouter.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+const UNSUBSCRIBE_LINK_WINDOW_SIZE = 60 // seconds
+const MAX_UNSUBSCRIBE_LINK_REQUEST_BY_WINDOW = 5
+
+const sender = { dv: 1, fingerprint: 'unsubscribe-link-handler' }
+const redisGuard = new RedisGuard()
+
+class UnsubscribeLinkRouter {
+    async init () {
+        const { keystone: context } = await getSchemaCtx('MultiPayment')
```
Comment (pahaz) [inline apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:21]: Use model related to this domain.

File: apps/condo/domains/notification/routes/unsubscribeLinkRouter.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        this.context = context
+    }
+
+
+    async checkLimits (req) {
+        const ip = req.ip.split(':').pop()
```
Comment (pahaz) [inline apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:27]: `req.ip` should not contains any ports.

File: apps/condo/domains/notification/constants/links.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+const UNSUBSCRIBE_LINK_PATH = '/unsubscribe'
```
Comment (pahaz) [inline apps/condo/domains/notification/constants/links.js:1]: we need to add `/api/` prefix or `/api/<domain>/` prefix

File: apps/condo/domains/notification/routes/unsubscribeLinkRouter.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                        isEnabled: false,
+                    })
+                }
+            }
+        } catch (e) {
+            return res.redirect('/500')
+        }
+
+        // redirect to unsubscribed page
+        return res.redirect('/unsubscribed')
```
Comment (pahaz) [inline apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:69]: Hardcode %) but you have a const

File: apps/condo/domains/notification/UnsubscribeMiddleware.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+const { UnsubscribeLinkRouter } = require('@condo/domains/notification/routes/unsubscribeLinkRouter')
+
+class UnsubscribeMiddleware {
+    async prepareMiddleware () {
+        // this route can not be used for csrf attack (because no cookies and tokens are used in a public route)
+        // nosemgrep: javascript.express.security.audit.express-check-csurf-middleware-usage.express-check-csurf-middleware-usage
+        const app = express()
+
+        const router = new UnsubscribeLinkRouter()
```
Comment (pahaz) [inline apps/condo/domains/notification/UnsubscribeMiddleware.js:12]: is it possible to move this router inside this file?

File: apps/condo/domains/notification/routes/unsubscribeLinkRouter.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        )
+    }
+
+    async handleRequest (req, res) {
+        try {
+            // check throttling limits
+            await this.checkLimits(req)
+
+            // get notification parameter
+            const { id } = req.query
```
Comment (pahaz) [inline apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:41]: check that `id` is not undefined!

File: apps/condo/domains/notification/routes/unsubscribeLinkRouter.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+            // get related message
+            const message = await Message.getOne(this.context, { id })
```
Comment (pahaz) [inline apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:44]: You also need to check access!!

Comment (pahaz) [reply apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:44]: Or check that the message is for anonymous user!

File: apps/condo/domains/notification/routes/unsubscribeLinkRouter.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                        sender,
+                        email: message.email,
+                        messageType: message.type,
+                        messageTransport: EMAIL_TRANSPORT,
+                        isEnabled: false,
+                    })
+                }
+            }
+        } catch (e) {
+            return res.redirect('/500')
```
Comment (pahaz) [inline apps/condo/domains/notification/routes/unsubscribeLinkRouter.js:65]: add logging like `logger.error({ err: e, msg: "handleUnsubscribe Error" })`

File: apps/condo/domains/notification/schema/NotificationAnonymousSetting.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        code: BAD_USER_INPUT,
+        type: NO_NEED_TO_ENABLE_NOTIFICATIONS,
+        message: 'No need to enable notifications. All notifications enabled by default. You may just delete this setting instead.',
+    },
+}
+
+const NotificationAnonymousSetting = new GQLListSchema('NotificationAnonymousSetting', {
+    schemaDoc: 'Anonymous contact notifications settings',
+    fields: {
+        email: {
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/NotificationAnonymousSetting.js:28]: may be we need to add phone?

File: apps/condo/domains/notification/utils/serverSchema/helpers.js [hunk 12]
```diff
@@ -4,14 +4,52 @@ const {
...
+
+/**
+ * @param context
+ * @param email
+ * @param messageType
+ * @returns {Promise<NotificationAnonymousSetting[]>}
+ */
+async function getAnonymousSettings (context, email, messageType) {
+    return await NotificationAnonymousSetting.getAll(context, {
+        email: email,
```
Comment (pahaz) [inline apps/condo/domains/notification/utils/serverSchema/helpers.js:43]: email or phone?

File: apps/condo/domains/notification/utils/serverSchema/helpers.js [hunk 13]
```diff
@@ -26,20 +64,12 @@ async function getUserSettingsForMessage (context, message) {
...
-        user: { id: message.user.id },
-        deletedAt: null,
-        OR: [
-            { messageType: null }, // possible settings for all messages
-            { messageType: message.type }, // settings for specific message type
-        ],
-    })
+    /** @type {NotificationUserSetting[] | NotificationAnonymousSetting[]} */
+    const notificationUserSettings = isAnonymous
+        ? await getAnonymousSettings(context, message.email, message.type)
```
Comment (pahaz) [inline apps/condo/domains/notification/utils/serverSchema/helpers.js:71]: + phone

File: apps/condo/domains/organization/access/SendNewBillingReceiptFilesNotificationsService.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Generated by `createservice organization.SendNewBillingReceiptFilesNotificationsService '--type=mutations'`
+ */
+const { throwAuthenticationError } = require('@open-condo/keystone/apolloErrorFormatter')
+
+async function canSendNewBillingReceiptFilesNotifications ({ authentication: { item: user } }) {
```
Comment (pahaz) [inline apps/condo/domains/organization/access/SendNewBillingReceiptFilesNotificationsService.js:6]: It's not a part of organization domain (or organization service). Move it to billing

File: apps/condo/domains/organization/schema/SendNewBillingReceiptFilesNotificationsService.test.js [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+                unitType: account.unitType,
+                isVerified: true,
+            })
+
+            // deleted receipt
+            await updateTestBillingReceipt(admin, billingReceipt.id, {
+                deletedAt: dayjs().toISOString(),
+            })
+
+            await wait(1000)
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/SendNewBillingReceiptFilesNotificationsService.test.js:236]: use waitFor

File: apps/condo/domains/organization/schema/SendNewBillingReceiptFilesNotificationsService.js [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+            access: true,
+            type: 'type SendNewBillingReceiptFilesNotificationsOutput { notificationsSent: Int! }',
+        },
+    ],
+
+    mutations: [
+        {
+            access: access.canSendNewBillingReceiptFilesNotifications,
+            schema: 'sendNewBillingReceiptFilesNotifications(data: SendNewBillingReceiptFilesNotificationsInput!): SendNewBillingReceiptFilesNotificationsOutput',
+            resolver: async (parent, args, context, info, extra = {}) => {
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/SendNewBillingReceiptFilesNotificationsService.js:112]: It looks like a part of another task! Why it's not a part of existing task? We already have a task for sending notification for new receipts.

It's better to move this logic close to that task! It should be similar logic

File: apps/condo/domains/notification/UnsubscribeMiddleware.js (hunk 1 above)
Comment (ekabardinsky) [reply apps/condo/domains/notification/UnsubscribeMiddleware.js:3]: I'll more constants there

File: apps/condo/domains/billing/tasks/sendNewBillingReceiptFilesNotifications.js [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+                    originalFilename,
+                })
+            }
+
+            return attachment
+        })
+}
+
+async function sendNewBillingReceiptFilesNotifications ({ organizationId, organizationName, sender, period, watermark }) {
+    const { keystone: context } = await getSchemaCtx('BillingReceiptFile')
```
Comment (dkoviazin) [inline apps/condo/domains/billing/tasks/sendNewBillingReceiptFilesNotifications.js:48]: it is not an async function

File: apps/condo/domains/billing/tasks/sendNewBillingReceiptFilesNotifications.js [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        const billingProperty = await BillingProperty.getOne(context, { id: billingPropertyId })
+
+        // check if no matching property exists
+        if (isNil(get(billingProperty, ['property', 'id']))) {
+            continue
+        }
+
+        // get contacts
+        const contacts = await Contact.getAll(context, {
+            property: { id: billingProperty.property.id },
```
Comment (dkoviazin) [inline apps/condo/domains/billing/tasks/sendNewBillingReceiptFilesNotifications.js:94]: you can use addressKey to match billingProperty with property

File: apps/condo/domains/notification/access/NotificationAnonymousSetting.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
+// @ts-nocheck
```
Comment (dkoviazin) [inline apps/condo/domains/notification/access/NotificationAnonymousSetting.js:1]: ts-nocheck can be removed

File: apps/condo/domains/notification/schema/NotificationAnonymousSetting.js [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+        code: BAD_USER_INPUT,
+        type: EMPTY_NOTIFICATION_ANONYMOUS_SETTING_FIELDS_ERROR,
+        message: 'One of the "phone" or "email" fields should be provided.',
+    },
+}
+
+const NotificationAnonymousSetting = new GQLListSchema('NotificationAnonymousSetting', {
+    schemaDoc: 'Anonymous contact notifications settings',
+    fields: {
+        email: {
```
Comment (dkoviazin) [inline apps/condo/domains/notification/schema/NotificationAnonymousSetting.js:34]: maybe, we need to normalize email and phone?

**Review by dkoviazin — APPROVED** [review]

**Pull request merged**